[READY] Apply chunks from bottom to top

micbou opened the pull request:
Without the proposed fix, the test included in that PR fails as follows
```
FAIL: ycm.tests.vimsupport_test.ReplaceChunksInBuffer_LineOverlappingChunks_test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "C:\Users\micbou\projects\YouCompleteMe\python\ycm\tests\vimsupport_test.py", line 768, in ReplaceChunksInBuffer_LineOverlappingChunks_test
    AssertBuffersAreEqualAsBytes( expected_buffer, result_buffer )
  File "C:\Users\micbou\projects\YouCompleteMe\python\ycm\tests\vimsupport_test.py", line 86, in AssertBuffersAreEqualAsBytes
    eq_( ToBytes( result_line ), ToBytes( expected_line ) )
AssertionError: b'    third line' != b'    third '
```
Found the issue while trying to add code formatting support to the language server completer. The test is based on a real-world scenario when formatting with jdt.ls (for some reason, jdt.ls is including the newline of the previous line when fixing indentation).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2898)
<!-- Reviewable:end -->

bstaletic commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Thanks for working on LSP formatting, I'm looking forward to this feature.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L4kO7UpqgdNy9t9hALs:b-nblsyd)*
<!-- Sent from Reviewable.io -->

micbou commented:
Found two additional bugs in the logic:
 - text is not correctly replaced when the chunk is `\n`. This is because the `splitlines` method splits the `\n` string into `[ '' ]` instead of `[ '', '' ]`. We replace that method with our own implementation (same as `SplitLines` in ycmd but for byte strings);
 - the character delta is added to the previous one even if the replacement text spans on multiple lines.

Added two tests that cover both issues and refactored the `ReplaceChunk` function so that we don't need to add the deltas outside that function.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L4lctdlJWkQK0qHpFj9:b-ue89y)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Reviewed 1 of 2 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/vimsupport.py, line 871 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L4oxBiZnCVY21EsfGqB:-L4oxBiZnCVY21EsfGqC:bjo3h7f) ([raw file](https://github.com/valloric/youcompleteme/blob/6b7ff33cf6a1b50d1f4a9c152ab1b14173a77620/python/ycm/vimsupport.py#L871)):*
> ```Python
>     } )
> 
>   line_delta += replacement_lines_count - source_lines_count
> ```

Why are we adding to the original `line_delta`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898)*
<!-- Sent from Reviewable.io -->

micbou commented:
Found another issue where the computation of the character delta was wrong. Added a test.

---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/vimsupport.py, line 871 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L4oxBiZnCVY21EsfGqB:-L4pGHa3nR3bxjyzMj4V:b-ekopnz) ([raw file](https://github.com/valloric/youcompleteme/blob/6b7ff33cf6a1b50d1f4a9c152ab1b14173a77620/python/ycm/vimsupport.py#L871)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Why are we adding to the original `line_delta`?
</blockquote></details>

We were previously [summing the deltas outside this function](https://github.com/Valloric/YouCompleteMe/blob/d0c6b6186e7a9ab7d5e770add3079200a518d908/python/ycm/vimsupport.py#L794-L795). It's more convenient to do it directly in the function. If you are asking why we are actually summing up the line deltas, it's because for each chunk that's adding or removing lines in the text, we need to shift all remaining chunks with that amount of lines.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L4pIkSpfDKmM-xfMnF7:bzf69no)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/vimsupport.py, line 853 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L4pKzY4dLeYzI9MLGL9:-L4pKzY5auDsn2IEeHHI:btn0igj) ([raw file](https://github.com/valloric/youcompleteme/blob/4cd73ab6fcfe8895e81343263d48dd6142d4cdf0/python/ycm/vimsupport.py#L853)):*
> ```Python
>     char_delta += end_replacement_column - ( end_column - start_column )
>   else:
>     char_delta = end_replacement_column - ( end[ 'column_num' ] - 1 )
> ```

Can we add a comment saying why `end_column` doesn't work here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/vimsupport.py, line 853 at r5](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L4pKzY4dLeYzI9MLGL9:-L4pd7whvNJ35-Mv2NVN:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/4cd73ab6fcfe8895e81343263d48dd6142d4cdf0/python/ycm/vimsupport.py#L853)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Can we add a comment saying why `end_column` doesn't work here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Once again, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 1 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L4t5BmL_m2eu_fsDvej:b-b02whk)*
<!-- Sent from Reviewable.io -->

micbou commented:
I've rewritten the logic to apply the chunks from the bottom to the top of the buffer. This simplifies a lot the code as we don't need to compute the deltas after applying each chunk. Some tests have been removed because they were testing those deltas computations.

---

Reviewed 2 of 2 files at r7, 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L4zsK9qdTd0sxEKVGI-:b-oa7pkn)*
<!-- Sent from Reviewable.io -->

puremourning commented:
I tested this by running the same set of tests i ran for the original implementations:

- FixIt_Clang_cpp11.cpp (in clang tests)
- file1.js in javascript tests (RefactorRename)
- FixItTestCase.cs (in cs tests)

However, i think i am seeing a regression with the java completer.

test case:

- open ycmd/tests/java/testdata/simple_eclipse_project/src/com/test/TestLauncher.java
- put the cursor on `Lauchable` on line 28
- run FixIt
- Expect: valid java generated
- Actual: it seems to mangle the output (the braces are no longer correct)

```java

  public static void main( String[] args ) {
    TestLauncher l = new TestLauncher( 10 );
    l.Run( new Launchable() {
      @Override
      public void launch() {
        AbstractTestWidget w = factory.getWidget( "Test" );
        w.doSomethingVaguelyUseful();

        System.out.println( "Did something useful: " + w.getWidgetInfo() );
      }
	}@Override
	public void launch(TestFactory f) {
		

	
    });
    static_method();
    TestLauncher t = new TestLauncher( 4 );
    t.Run( null );
  }


``` 

can you confirm/deny this behaviours?

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/vimsupport.py, line 715 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L5-bMMZJAJO_dewM4Qd:-L5-bMMZJAJO_dewM4Qe:bti31hq) ([raw file](https://github.com/valloric/youcompleteme/blob/d7022bdb4fdf15b5f77aacfbe7303525acadd65b/python/ycm/vimsupport.py#L715)):*
> ```Python
> 
>   # We apply the edits file-wise for efficiency, and because we must track the
>   # file-wise offset deltas (caused by the modifications to the text).
> ```

we no longer maintain deltas, thought the reasoning applies the same.

---

*[python/ycm/vimsupport.py, line 768 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L5-boGEPZCVkqKjmspl:-L5-boGEPZCVkqKjmspm:b-smc3l2) ([raw file](https://github.com/valloric/youcompleteme/blob/d7022bdb4fdf15b5f77aacfbe7303525acadd65b/python/ycm/vimsupport.py#L768)):*
> ```Python
>   # We apply the chunks from the bottom of the buffer to its top so that we
>   # don't need to adjust the position of the remaining chunks due to text
>   # changes. This assumes that chunks are not overlapping.
> ```

are we sure that this non-overlappingness is true for existing fixit scenarios? When i implmented fixits for clang i tested every one in the test file from clang source at the time of writing.

perhaps we should do some due diligence and update the ycmd api to say that we don't allow edits to overlap. tbh we probably didn't before and if it worked it was more luck than judgement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L5-dGSFlpwh0t2dXKHd:b-32odkj)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/vimsupport.py, line 772 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L5-qMYxyTlgDzmmI4MI:-L5-qMYybVENmTg59GON:b-m8f3ox) ([raw file](https://github.com/valloric/youcompleteme/blob/d7022bdb4fdf15b5f77aacfbe7303525acadd65b/python/ycm/vimsupport.py#L772)):*
> ```Python
>     chunk[ 'range' ][ 'start' ][ 'line_num' ],
>     chunk[ 'range' ][ 'start' ][ 'column_num' ]
>   ), reverse = True )
> ```

as discussed, this sort is _stable_, which is annoyingly not quite what we want.

Changes must be applied strictly in reverse order of location, but where locations are the _same_, they must be applied in reverse order of appearance in the list.

therefore, we need to fully reverse the list, _then_ perform the stable sort on location (with reverse=True, to sort by later edits first, maintaining the - now reversed - sequence of items with the same key)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/vimsupport.py, line 715 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L5-bMMZJAJO_dewM4Qd:-L506NNjgTIKaEqnkoH0:b-ddo8jc) ([raw file](https://github.com/valloric/youcompleteme/blob/d7022bdb4fdf15b5f77aacfbe7303525acadd65b/python/ycm/vimsupport.py#L715)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

we no longer maintain deltas, thought the reasoning applies the same.
</blockquote></details>

Removed the second part of the sentence.

---

*[python/ycm/vimsupport.py, line 768 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L5-boGEPZCVkqKjmspl:-L50KZKmyfn5SC21ZzAQ:b-prwqyd) ([raw file](https://github.com/valloric/youcompleteme/blob/d7022bdb4fdf15b5f77aacfbe7303525acadd65b/python/ycm/vimsupport.py#L768)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

are we sure that this non-overlappingness is true for existing fixit scenarios? When i implmented fixits for clang i tested every one in the test file from clang source at the time of writing.

perhaps we should do some due diligence and update the ycmd api to say that we don't allow edits to overlap. tbh we probably didn't before and if it worked it was more luck than judgement.
</blockquote></details>

Yes, this should be mentioned in the API.

---

*[python/ycm/vimsupport.py, line 772 at r9](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-L5-qMYxyTlgDzmmI4MI:-L506ZTrvfyt29j4fA4J:bg1n70i) ([raw file](https://github.com/valloric/youcompleteme/blob/d7022bdb4fdf15b5f77aacfbe7303525acadd65b/python/ycm/vimsupport.py#L772)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

as discussed, this sort is _stable_, which is annoyingly not quite what we want.

Changes must be applied strictly in reverse order of location, but where locations are the _same_, they must be applied in reverse order of appearance in the list.

therefore, we need to fully reverse the list, _then_ perform the stable sort on location (with reverse=True, to sort by later edits first, maintaining the - now reversed - sequence of items with the same key)
</blockquote></details>

Done. Added a test that fails without this change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

confirmed this works with the language server fixits.

thanks!

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L50N61wyw6UTeLhbbkP:bx0t4z1)*
<!-- Sent from Reviewable.io -->

micbou commented:
@zzbot r+

---

Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2898#-:-L510AySWNHUkKlUmhMI:bi59kwd)*
<!-- Sent from Reviewable.io -->
